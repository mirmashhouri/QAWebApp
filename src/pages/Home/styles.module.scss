.home {
  margin-bottom: 150px;
}
