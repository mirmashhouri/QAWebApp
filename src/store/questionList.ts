import { PayloadAction, createSlice } from "@reduxjs/toolkit";

import { Question, getQuestionList } from "../services/questionService";
import { AppThunk, AppState } from ".";

interface QuestionList {
  readyStatus: string;
  items: Question[];
  error: string | null;
}

export const initialState: QuestionList = {
  readyStatus: "invalid",
  items: [],
  error: null,
};

interface Edit {
  oldData: Question;
  newData: Question;
}
const questionList = createSlice({
  name: "questionList",
  initialState,
  reducers: {
    getRequesting: (state: QuestionList) => {
      state.readyStatus = "request";
    },
    getSuccess: (state, { payload }: PayloadAction<Question[]>) => {
      state.readyStatus = "success";
      state.items = payload;
    },
    getFailure: (state, { payload }: PayloadAction<string>) => {
      state.readyStatus = "failure";
      state.error = payload;
    },
    insert: (state, { payload }: PayloadAction<Question>) => {
      state.readyStatus = "success";
      state.items.push(payload);
    },
    deleteOne: (state, { payload }: PayloadAction<Question>) => {
      state.readyStatus = "success";
      state.items = state.items.filter((q) => q.question !== payload.question);
    },
    deleteAll: (state: QuestionList) => {
      state.readyStatus = "success";
      state.items = [];
    },
    edit: (state, { payload }: PayloadAction<Edit>) => {
      state.readyStatus = "success";
      const index = state.items.findIndex(
        (q) => q.question === payload.oldData.question
      );
      state.items[index] = payload.newData;
    },
    sort: (state: QuestionList) => {
      state.readyStatus = "success";
      state.items.sort((a, b) =>
        a.question.toLowerCase().localeCompare(b.question.toLowerCase())
      );
    },
  },
});

export default questionList.reducer;
export const {
  getRequesting,
  getSuccess,
  getFailure,
  insert,
  deleteAll,
  sort,
  deleteOne,
  edit,
} = questionList.actions;

export const fetchQuestionList = (): AppThunk => async (dispatch) => {
  dispatch(getRequesting());

  const { error, data } = await getQuestionList();

  if (error) {
    dispatch(getFailure(error.message));
  } else {
    dispatch(getSuccess(data as Question[]));
  }
};
export const addNewQuestion =
  (data: Question, delay: number): AppThunk =>
  async (dispatch) => {
    setTimeout(() => {
      dispatch(getRequesting());
      dispatch(insert(data));
    }, delay);
  };
export const deleteAllQuestions = (): AppThunk => async (dispatch) => {
  dispatch(getRequesting());
  dispatch(deleteAll());
};
export const deleteQuestion =
  (data: Question): AppThunk =>
  async (dispatch) => {
    dispatch(getRequesting());
    dispatch(deleteOne(data));
  };
export const editQuestion =
  (oldData: Question, newData: Question, delay: number): AppThunk =>
  async (dispatch) => {
    setTimeout(() => {
      dispatch(getRequesting());
      const editData: Edit = {
        oldData,
        newData,
      };
      dispatch(edit(editData));
    }, delay);
  };
export const sortQuestions = (): AppThunk => async (dispatch) => {
  dispatch(getRequesting());
  dispatch(sort());
};

const shouldFetchQuestionList = (state: AppState) =>
  state.questionList.readyStatus !== "success";

export const fetchQuestionListIfNeed = (): AppThunk => (dispatch, getState) => {
  if (shouldFetchQuestionList(getState())) return dispatch(fetchQuestionList());

  return null;
};
