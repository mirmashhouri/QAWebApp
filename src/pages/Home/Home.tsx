import { FC, useEffect, memo, useState } from "react";
import { RouteComponentProps } from "react-router-dom";
import { useDispatch, useSelector, shallowEqual } from "react-redux";
import { Helmet } from "react-helmet";
import { AppState, AppThunk } from "../../store";
import {
  fetchQuestionListIfNeed,
  addNewQuestion,
  deleteAllQuestions,
  sortQuestions,
  deleteQuestion,
  editQuestion,
} from "../../store/questionList";
import { Question, QuestionForm } from "../../services/questionService";
import { Items, Form, Tooltip } from "../../components";
import styles from "./styles.module.scss";

export type Props = RouteComponentProps;

const Home: FC<Props> = (): JSX.Element => {
  const dispatch = useDispatch();
  const { readyStatus, items } = useSelector(
    ({ questionList }: AppState) => questionList,
    shallowEqual
  );

  const initialForm: QuestionForm = {
    isEdit: false,
    data: undefined,
  };
  const [editIndex, setEditIndex] = useState(1);
  const [isVisible, setIsVisible] = useState(false);
  const [formData, setFormData] = useState(initialForm);
  const FormLoad = (data?: Question) => {
    setIsVisible(false);
    setEditIndex(editIndex + 1);
    setFormData(data ? { isEdit: true, data } : initialForm);
    setIsVisible(true);
  };
  // Fetch client-side data here
  useEffect(() => {
    dispatch(fetchQuestionListIfNeed());
  }, [dispatch]);

  const AddOrEdit = (data: Question, hasDelay: boolean) => {
    const delay = hasDelay ? 5000 : 0;
    if (formData.isEdit && formData.data)
      dispatch(editQuestion(formData.data, data, delay));
    else {
      dispatch(addNewQuestion(data, delay));
    }
    setIsVisible(false);
  };
  const DeleteAllOnClick = () => {
    dispatch(deleteAllQuestions());
  };
  const SortOnClick = () => {
    dispatch(sortQuestions());
  };
  const OnDelete = (data: Question) => {
    dispatch(deleteQuestion(data));
  };
  const OnEdit = (data: Question) => {
    FormLoad(data);
  };
  const renderList = () => {
    if (!readyStatus || readyStatus === "invalid" || readyStatus === "request")
      return <p>Loading...</p>;

    if (readyStatus === "failure") return <p>Loading</p>;

    return (
      <div className="container-md">
        <Items items={items} onDelete={OnDelete} onEdit={OnEdit} />
        {isVisible ? (
          <Form key={editIndex} onAddOrEdit={AddOrEdit} item={formData} />
        ) : null}
        <Tooltip content="Open the question form to insert new">
          <button
            className="btn btn-light m-1"
            type="button"
            onClick={() => FormLoad(undefined)}
          >
            Add New
          </button>
        </Tooltip>
        <Tooltip content="Sort alphabeta all questions">
          <button
            className="btn btn-light m-1"
            type="button"
            onClick={SortOnClick}
          >
            Sort
          </button>
        </Tooltip>
        <Tooltip content="Dell all questions">
          <button
            className="btn btn-danger m-1"
            type="button"
            onClick={DeleteAllOnClick}
          >
            Delete All
          </button>
        </Tooltip>
      </div>
    );
  };

  return (
    <div className={styles.home}>
      <Helmet title="Home page" />
      {renderList()}
    </div>
  );
};

// Fetch server-side data here
export const loadData = (): AppThunk[] => [
  fetchQuestionListIfNeed(),
  // More pre-fetched actions...
];

export default memo(Home);
