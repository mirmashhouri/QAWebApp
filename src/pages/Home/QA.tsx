import { FC, useEffect, memo } from "react";
import { RouteComponentProps } from "react-router-dom";
import { useDispatch, useSelector, shallowEqual } from "react-redux";
import { Helmet } from "react-helmet";

import { AppState, AppThunk } from "../../store";
import {
  fetchQuestionListIfNeed,
  addNewQuestion,
} from "../../store/questionList";
import { Question } from "../../services/questionService";
import { Items, Form } from "../../components";
import styles from "./styles.module.scss";

export type Props = RouteComponentProps;

const QA: FC<Props> = (): JSX.Element => {
  const dispatch = useDispatch();
  const { readyStatus, items } = useSelector(
    ({ questionList }: AppState) => questionList,
    shallowEqual
  );

  // Fetch client-side data here
  useEffect(() => {
    dispatch(fetchQuestionListIfNeed());
  }, [dispatch]);

  const addNew = (data: Question) => {
    dispatch(addNewQuestion(data));
  };
  const renderList = () => {
    if (!readyStatus || readyStatus === "invalid" || readyStatus === "request")
      return <p>Loading...</p>;

    if (readyStatus === "failure") return <p>Oops, Failed to load list!</p>;

    return (
      <div>
        <Items items={items} />
        <Form onAdd={addNew} />
      </div>
    );
  };

  return (
    <div className={styles.Home}>
      <Helmet title="Q/A page" />
      {renderList()}
    </div>
  );
};

// Fetch server-side data here
export const loadData = (): AppThunk[] => [
  fetchQuestionListIfNeed(),
  // More pre-fetched actions...
];

export default memo(QA);
