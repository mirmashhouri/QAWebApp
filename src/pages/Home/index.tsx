import loadable from "@loadable/component";

import { Loading, ErrorBoundary } from "../../components";
import { Props, loadData } from "./QA";

const QA = loadable(() => import("./QA"), {
  fallback: <Loading />,
});

export default (props: Props): JSX.Element => (
  <ErrorBoundary>
    <QA {...props} />
  </ErrorBoundary>
);
export { loadData };
