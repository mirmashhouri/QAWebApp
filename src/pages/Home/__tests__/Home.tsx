import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

import { fetchQuestionListIfNeed } from "../../../store/questionList";
import mockStore from "../../../utils/mockStore";
import Home from "../Home";

describe("<Home />", () => {
  const renderHelper = (reducer = { readyStatus: "invalid" }) => {
    const { dispatch, ProviderWithStore } = mockStore({
      questionList: reducer,
    });
    const { container } = render(
      <ProviderWithStore>
        <MemoryRouter>
          {/*
            @ts-expect-error */}
          <Home />
        </MemoryRouter>
      </ProviderWithStore>
    );

    return { dispatch, firstChild: container.firstChild };
  };

  it("should fetch data when page loaded", () => {
    const { dispatch } = renderHelper();

    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch.mock.calls[0][0].toString()).toBe(
      fetchQuestionListIfNeed().toString()
    );
  });

  it("renders the loading status if data invalid", () => {
    expect(renderHelper().firstChild).toMatchSnapshot();
  });

  it("renders the loading status if requesting data", () => {
    const reducer = { readyStatus: "request" };

    expect(renderHelper(reducer).firstChild).toMatchSnapshot();
  });

  it("renders an error if loading failed", () => {
    const reducer = { readyStatus: "failure" };

    expect(renderHelper(reducer).firstChild).toMatchSnapshot();
  });

  it("renders the <Items /> if loading was successful", () => {
    const reducer = {
      readyStatus: "success",
      items: [
        {
          question: "What is your name?",
          answer: "My name is Abbas",
        },
      ],
    };

    expect(renderHelper(reducer).firstChild).toMatchSnapshot();
  });
});
