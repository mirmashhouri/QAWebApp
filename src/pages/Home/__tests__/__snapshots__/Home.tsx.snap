// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> renders an error if loading failed 1`] = `
<div
  class="home"
>
  <p>
    Loading
  </p>
</div>
`;

exports[`<Home /> renders the <Items /> if loading was successful 1`] = `
<div
  class="home"
>
  <div
    class="container-md"
  >
    <div
      class="items"
    >
      <div
        class="tooltip-wrapper"
      >
        <h4>
          Question List
        </h4>
      </div>
      <ul>
        <div
          class="item"
        >
          <ul>
            <li>
              <div
                class="row"
              >
                <div
                  class="col-7 col-md-9"
                >
                  <button
                    class="w-100 text-left"
                    type="button"
                  >
                    What is your name?
                  </button>
                  <div
                    class="answer "
                  >
                    My name is Abbas
                  </div>
                </div>
                <div
                  class="col-5 col-md-3 action"
                >
                  <button
                    type="button"
                  >
                    Edit
                  </button>
                  <span>
                     | 
                  </span>
                  <button
                    type="button"
                  >
                    Delete
                  </button>
                </div>
              </div>
            </li>
          </ul>
        </div>
      </ul>
    </div>
    <div
      class="tooltip-wrapper"
    >
      <button
        class="btn btn-light m-1"
        type="button"
      >
        Add New
      </button>
    </div>
    <div
      class="tooltip-wrapper"
    >
      <button
        class="btn btn-light m-1"
        type="button"
      >
        Sort
      </button>
    </div>
    <div
      class="tooltip-wrapper"
    >
      <button
        class="btn btn-danger m-1"
        type="button"
      >
        Delete All
      </button>
    </div>
  </div>
</div>
`;

exports[`<Home /> renders the loading status if data invalid 1`] = `
<div
  class="home"
>
  <p>
    Loading...
  </p>
</div>
`;

exports[`<Home /> renders the loading status if requesting data 1`] = `
<div
  class="home"
>
  <p>
    Loading...
  </p>
</div>
`;
